### PR TITLE
Fix TRAMP issues

### LINKED ISSUE #523: projectile-global-mode prevents new processes from working over TRAMP
I don't have time to investiguate it right now but I should be able to this week.

### PR DESCRIPTION
**Pull request by Silex**
This should fix #523 and #835.

Don't merge this yet (it lacks changelogs, etc). Right now the idea is for people having the same problem to test if my fix actually fixes all scenarios.

The idea is pretty simple: only resolve project-root for local files, or when the remote files are _connected_.

So far this seems to resolve all the issues for me, but I only tested emacs `25.1`.

Things to test:

- `/scpx:user@host:somefile`
- `/sudo:root@localhost:somefile`

Note: by me, the issues only happened when `projectile-rails-global-mode` was turned on. Projectile alone worked pretty well. This is because `projectile-rails-mode` uses `projectile` to determine wether it'd turn on or not, and this was just too much for TRAMP.

-----------------

- [x] The commits are consistent with our [contribution guidelines](../CONTRIBUTING.md)
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [x] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)

### CODE REVIEW

**Comment by Silex** [comment]
@timvisher, @2mc, @tonycpsu, @KGZM, @ackerleytng, @IvanMalison, @spaceotter, @stardiviner, @puffnfresh, @seagle0128, @bhrgunatha, @binarin: please test and report.

**Comment by timvisher** [comment]
Whoa! Nice! :)

I'll try to test this out today!

File: projectile.el [hunk 1]
```diff
@@ -863,17 +863,20 @@ topmost sequence of matched directories.  Nil otherwise."
...
-                  (cache-value (gethash cache-key projectile-project-root-cache)))
-             (if (and cache-value (file-exists-p cache-value))
-                 cache-value
-               (let ((value (funcall func (file-truename dir))))
-                 (puthash cache-key value projectile-project-root-cache)
-                 value))))
-         projectile-project-root-files-functions)
+  (let* ((dir default-directory)
+         (is-local (not (file-remote-p dir)))
+         (is-connected (file-remote-p default-directory nil t)))
```
Comment (bbatsov) [inline projectile.el]: What doest this mean?

Comment (Silex) [inline projectile.el]: huh? `is-local` is true if the file is local, `is-connected` is true if the file is remote but we are already connected.

Comment (Silex) [inline projectile.el]: Ah, the second `default-directory` should be `dir`, I'll make a fixup commit

**Comment by bhrgunatha** [comment]
I'm on emacs 25.1 now - tramp is working for me now with projectile-mode (also looks OK with projectile-rails-mode too)

**Comment by Silex** [comment]
@bhrgunatha: with or without my patch? Also, to be exact it's `projectile-rails-global-mode` that triggers the `projectile` bug, both with `/sudo:` and `/scpx:`.

**Comment by timvisher** [comment]
```
$ git log -1
commit 6dd3cca8c5d3b321c9c206967cf32985103eff73
Author: Philippe Vaucher <philippe.vaucher@stvs.com>
Date:   Thu Mar 23 13:39:15 2017 +0100

    fixup! Fix TRAMP issues
```

```
GNU Emacs 25.1.2 (x86_64-apple-darwin15.6.0) of 2016-09-19
```

Cabs running fine… Purring like a kitten…

(Supergreen)

(Meaning it works great for me :) )

File: projectile.el (hunk 1 above)
Comment (Silex) [inline projectile.el]: @bbatsov: I don't really understand your original question, can you clarify?

**Comment by tonycpsu** [comment]
Looks good here on Emacs 24.5.

**Comment by bhrgunatha** [comment]
@Silex 
Sorry I wasn't clear and I'm afraid this won't be very helpful either.
Without any patches applied after moving to 25.1, tramp is working without any other direct intervention needed from me.

**Comment by Silex** [comment]
Okay, I discovered that it's possible to update TRAMP itself. I'm building a test matrix based on:

- Projectile with or without my fix
- Emacs 24.5 or 25.1
- Projectile-Rails on or off
- TRAMP default or latest version

I'll report when it's done.

**Comment by seagle0128** [comment]
Works well on Emacs 25.1.

**Comment by Silex** [comment]
Okay, here are my tests so far:

| Version | Projectile branch | Projectile Rails? | TRAMP latest? | Failed |
|---------|-------------------|-------------------|---------------|--------|
|    24.5 | master            | no                | no            |      0 |
|    24.5 | master            | no                | yes           |      0 |
|    24.5 | master            | yes               | no            |      1 |
|    24.5 | master            | yes               | yes           |      1 |
|    24.5 | repair-tramp      | no                | no            |      0 |
|    24.5 | repair-tramp      | no                | yes           |      0 |
|    24.5 | repair-tramp      | yes               | no            |      1 |
|    24.5 | repair-tramp      | yes               | yes           |      0 |
|    25.1 | master            | no                | no            |      0 |
|    25.1 | master            | no                | yes           |      0 |
|    25.1 | master            | yes               | no            |      1 |
|    25.1 | master            | yes               | yes           |      1 |
|    25.1 | repair-tramp      | no                | no            |      0 |
|    25.1 | repair-tramp      | no                | yes           |      0 |
|    25.1 | repair-tramp      | yes               | no            |      0 |
|    25.1 | repair-tramp      | yes               | yes           |      0 |

We see that my branch repairs 3 tests out of 4, and that by updating TRAMP to latest version we get the last failing test. The tests were a bit of a pain to run, and I doubt they can actually become part of projectile's test suite due to the heavy setup required. Maybe a second travis-ci just for them might work tho. For the brave who like to look at work in progress, look at https://github.com/Silex/projectile/tree/tramp-tests/bugs

**Comment by Silex** [comment]
@bbatsov: I'd like to hear from you. I don't think I'll progress much more on this issue.

Here is the current status:

- I don't think my patch can really hurt anyone. Maybe performances a bit?
- More than one person reported it fixed the problem for them.
- Writing tests for this is complicated, I'm not sure it's automatable.  Also as I said I'm not even sure you'd want these tests inside the test suite, so I think I'll stop working on this.

For information, here is the latest update with the TRAMP guys: http://lists.gnu.org/archive/html/bug-gnu-emacs/2017-03/msg00979.html

Basically they say it makes sense to test for what I test, but that there's not much they can do. Emacs 26 might fix the issue, but so far with my tests the bug is still present.

**Comment by Silex** [comment]
@bbatsov: ping :-)

**Comment by bbatsov** [comment]
I'm extremely busy, so I don't really have time to go through the linked issues. What exactly is this supposed to fix. Frankly, I'm getting a bit annoyed from having to complicate the code base more and more and more on behalf of TRAMP, which I don't even use. I still don't get while someone would be developing over TRAMP...

So, briefly, what is this supposed to fix? 

P.S. Just looking at the patch I find it extremely odd there's a branch in the logic where we don't even attempt the find the project root.

**Comment by Silex** [comment]
> So, briefly, what is this supposed to fix?

It fixes the behavior where Emacs hangs because of projectile when you open a file over TRAMP. When Emacs hangs, very often the only solution is to kill emacs from another terminal.

> P.S. Just looking at the patch I find it extremely odd there's a branch in the logic where we don't even attempt the find the project root

This is exactly what the fix does: don't try to find the project root for remote files, if we are not connected to the remote host yet.

Here is what currently happens:

1. Open a file
2. Buffer is displayed "empty" (TRAMP starts connecting)
3. Projectile asks questions about the remote files (way too soon)
4. TRAMP is confused and hangs

Here is what my fix does:

1. Open a file
2. Buffer is displayed "empty" (TRAMP starts connecting)
3. Projectile asks questions about the remote files (gets told that no project root exists)
4. TRAMP eventually connects
5. File is displayed
6. Projectile asks questions about the remote files (now that it is connected it gets the right answer)

File: projectile.el [hunk 2]
```diff
@@ -863,17 +863,20 @@ topmost sequence of matched directories.  Nil otherwise."
...
-           (let* ((cache-key (format "%s-%s" func dir))
-                  (cache-value (gethash cache-key projectile-project-root-cache)))
-             (if (and cache-value (file-exists-p cache-value))
-                 cache-value
-               (let ((value (funcall func (file-truename dir))))
-                 (puthash cache-key value projectile-project-root-cache)
-                 value))))
-         projectile-project-root-files-functions)
+  (let* ((dir default-directory)
+         (is-local (not (file-remote-p dir)))
```
Comment (bbatsov) [inline projectile.el]: I'd add what you answered to me in the issue here as a comment. Otherwise people are going to have a hard time figuring out what does this do.

Comment (Silex) [inline projectile.el]: Good idea! Doing it now.

**Comment by bbatsov** [comment]
Apart from my small remark - this needs a changelog entry.

**Comment by Silex** [comment]
Does the changelog entry and the fixup commit that adds comments work for you?

I'm not sure how detailed the comments needs to be.

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@
 
 ### Changes
 
+* [#1129](https://github.com/bbatsov/projectile/pull/1129) Fix TRAMP issues.
```
Comment (bbatsov) [inline CHANGELOG.md]: There's a missing `:` here.

File: projectile.el [hunk 4]
```diff
@@ -863,17 +863,23 @@ topmost sequence of matched directories.  Nil otherwise."
...
-           (let* ((cache-key (format "%s-%s" func dir))
-                  (cache-value (gethash cache-key projectile-project-root-cache)))
-             (if (and cache-value (file-exists-p cache-value))
-                 cache-value
-               (let ((value (funcall func (file-truename dir))))
-                 (puthash cache-key value projectile-project-root-cache)
-                 value))))
-         projectile-project-root-files-functions)
+  ;; The `is-local' and `is-connected' variables are used to fix the behavior where Emacs hangs
+  ;; because of Projectile when you open a file over TRAMP. It basically prevents projectile from
```
Comment (bbatsov) [inline projectile.el]: projectile -> Projectile

File: projectile.el [hunk 5]
```diff
@@ -863,17 +863,23 @@ topmost sequence of matched directories.  Nil otherwise."
...
+  ;; trying to find information about files for which it's not possible to get informations from.
```
Comment (bbatsov) [inline projectile.el]: informations from -> information

**Comment by bbatsov** [comment]
Added a couple of tiny remark - squash and rebase after you apply them.

**Comment by Silex** [comment]
Alright, reworded a little bit more, this should be good to go.

**Comment by Silex** [comment]
:+1:

**Comment by timvisher** [comment]
dance cash

@bbatsov The biggest reason to develop over tramp is to develop on ephemeral resources like vagrant boxes or ec2 instances without loosing any of your context should you need to replace the ephemeral resource.
